File: api/facts.ts
export async function handler(event: any) {
  const topics = [
    "artificial intelligence", "robotics", "quantum physics", "cybersecurity", "smart devices",
    "machine learning", "virtual reality", "blockchain", "future tech", "space exploration",
    "nature", "animals", "ocean", "trees", "plants", "gardening",
    "sustainable energy", "time travel", "memes", "deep sea creatures", "neural networks",
    "future trends", "fashion", "sports", "pop culture", "movies",
    "self-help", "productivity", "office structure", "focus", "cooking"
  ];

  const shuffledTopics = topics.sort(() => 0.5 - Math.random()).slice(0, 20);

  try {
    const factPromises = shuffledTopics.map(async (topic) => {
      const prompt = `Give me a unique, clever, short, and lighthearted fact. Make it about ${topic}.`;

      const response = await fetch("https://api.openai.com/v1/chat/completions", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${process.env.OPENAI_API_KEY}`,
        },
        body: JSON.stringify({
          model: "gpt-4o-mini",
          messages: [{ role: "user", content: prompt }],
          max_tokens: 150,
          temperature: 1.2,
        }),
      });

      const data = await response.json();
      return data.choices?.[0]?.message?.content?.trim();
    });

    const facts = await Promise.all(factPromises);
    return {
      statusCode: 200,
      body: JSON.stringify(facts.filter(Boolean)),
    };
  } catch (error: any) {
    return {
      statusCode: 500,
      body: JSON.stringify({ error: error.message }),
    };
  }
}

File: dev/facts.py
from fastapi import FastAPI
from fastapi.responses import JSONResponse
import os
import random
import asyncio
import httpx
from dotenv import load_dotenv

load_dotenv()

app = FastAPI()

topics = [
    "artificial intelligence", "robotics", "quantum physics", "cybersecurity", "smart devices",
    "machine learning", "virtual reality", "blockchain", "future tech", "space exploration",
    "nature", "animals", "ocean", "trees", "plants", "gardening",
    "sustainable energy", "time travel", "memes", "deep sea creatures", "neural networks",
    "future trends", "fashion", "sports", "pop culture", "movies",
    "self-help", "productivity", "office structure", "focus", "cooking"
]

OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")


async def fetch_fact(client: httpx.AsyncClient, topic: str) -> str:
    prompt = f"Give me a unique, clever, short, and lighthearted fact. Make it about {topic}."

    try:
        response = await client.post(
            "https://api.openai.com/v1/chat/completions",
            headers={
                "Authorization": f"Bearer {OPENAI_API_KEY}",
                "Content-Type": "application/json"
            },
            json={
                "model": "gpt-4o-mini",
                "messages": [{"role": "user", "content": prompt}],
                "max_tokens": 150,
                "temperature": 1.2,
            }
        )
        response.raise_for_status()
        data = response.json()
        return data["choices"][0]["message"]["content"].strip()
    except Exception as e:
        print(f"Error fetching fact for {topic}: {e}")
        return None


@app.get("/facts")
async def get_facts():
    selected_topics = random.sample(topics, 20)

    async with httpx.AsyncClient() as client:
        facts = await asyncio.gather(*(fetch_fact(client, topic) for topic in selected_topics))

    filtered_facts = [fact for fact in facts if fact]
    return JSONResponse(content=filtered_facts)

if __name__ == "__main__":
    import uvicorn
    uvicorn.run("facts:app", host="127.0.0.1", port=8000, reload=True)

